herd/libdir/aarch64hwreqs.cat

```diff
@@ -56,104 +56,124 @@ let tc-ib = if "MTE" then [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & W
 let f-ib =
   [Imp & Instr & R]; iico_data; [B]; iico_ctrl; [_]
 
-(* HW TTD Updates permitted only for the AF or DB *)
-let HU = Imp & TTD & W
-assert empty HU \ (AF | DB)
-
 (* DSB-ordered-before *)
 let DSB-ob =
- [M | DC.CVAU | IC]; po; [dsb.full]; po; [~(Imp & M)]
+[M | DC.CVAU | IC]; po; [dsb.full]; po; [~(Imp & M) | Imp & Tag & R]
  | (if "ETS2" then [M | DC.CVAU | IC]; po; [dsb.full]; po; [Imp & TTD & M] else 0)
- | [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [~(Imp & M)]
+ | [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [~(Imp & M)| Imp & Tag & R]
  | (if "ETS2" then [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [Imp & TTD & M] else 0)
- | [Exp & W]; po; [dsb.st]; po; [~(Imp & M)]
+ | [Exp & W]; po; [dsb.st]; po; [~(Imp & M) | Imp & Tag & R]
  | (if "ETS2" then [Exp & W]; po; [dsb.st]; po; [Imp & TTD & M] else 0)
 
 (* CSE-ordered-before *)
 let EXC-ENTRY-CSE = EXC-ENTRY
 let EXC-RET-CSE = if not "ExS" || "EOS" then EXC-RET else {}
 let CSE = ISB | EXC-ENTRY-CSE | EXC-RET-CSE
+
+let Sync-EXC = EXC-ENTRY | EXC-RET
+
 let CSE-ob =
  [Exp & R]; ctrl; [CSE]; po
+ | [Exp & R]; pick-ctrl-dep; [CSE]; po
+ | [Exp & R]; addr; [Exp & M]; po; [CSE]; po
+ | [Exp & R]; pick-addr-dep; [Exp & M]; po; [CSE]; po
  | DSB-ob; [CSE]; po
+ | [Imp & Instr & R]; f-ib; [CSE]; po
  | [Imp & TTD & R]; tr-ib; [CSE]; po
  | [Imp & TTD & R]; tr-ib; po; [CSE]; po
  | [Imp & Tag & R]; tc-ib; [CSE]; po
- | [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & M) \ Tag) | TagCheck & FAULT]; po; [CSE]; po
- | [Imp & Instr & R]; f-ib; [CSE]; po
+ | [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & M) \ Tag) | TagCheck & FAULT & Sync-EXC]; po; [CSE]; po
```

Comment:
I think you don't need the `\ Tag`, LDG, STG, etc are unchecked instructions. Also I think, this part of the clause:
```
 | [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [TagCheck & FAULT & Sync-EXC]; po; [CSE]; po
```
is already included in
```
| [Imp & Tag & R]; tc-ib; [CSE]; po
```
so I think you can simplify this to
```
[Imp & Tag & R]; iico_data; [B]; iico_ctrl; [Exp & M]; po; [CSE]; po
```
or even
```
[Imp & Tag & R]; po; [CSE]; po
```
Can't you?

Reply:
I like the second-to-last simplification. It is a bit sneaky, in that it relies on the instruction semantics in async/asym mode not to draw the `iico_ctrl` edge to the memory effect, and on `CSE` effects being synchronous exceptions. However, those are the matters of definition and strike me as completely adequate. I'd like to use this phrasing.

----

I like the very last suggestion too, as it looks very clean, but it does not allow for discriminating between sync/async/asym fault handling mode, does it? Not every Implicit Tag Read enjoys this order, but only the one that leads to a synchronous Tag Check fault exception. I suppose, precisely that can be defined as a filter on `Imp & Tag & R` rather than on Fault effects. Namely, we could do:
```
let MayRaiseSyncExc = domain([Imp & Tag & R]; iico_data; [B]; iico_ctrl; [Exp & M | TagCheck & EXC-ENTRY])
```
and adjust your last suggestion to
```
[Imp & Tag & R & MayRaiseSyncExc]; po; [CSE]; po
```

To attach this point to a litmus test, consider this:
```
AArch64 MP+dmb.stpt+tcib-po-isb-po--async
Variant=memtag,async
{
0:X1=x; 0:X3=y:green; 0:X2=y:red;
1:X1=x; 1:X3=y:red; 1:X2=1;
}
P0          | P1          ;
MOV W0,#1   | LDR W2,[X3] ;
STR W0,[X1] |L0:          ;
DMB ST      | ISB         ;
STG X2,[X3] | LDR W0,[X1] ;
exists ~Fault(P1,y) /\ 1:X0=0
```
The last suggestion would forbid this, although it only needs to be forbidden in sync mode. (And async mode, I guess.)

Reply:
Makes sense, thanks!

In that case, I think I vote for:
```
[Imp & Tag & R]; iico_data; [B]; iico_ctrl; [Exp & M]; po; [CSE]; po
```
and I think it might be easier to transliterate but it's up to you. And you can also add a comment why this works (ie., in async there is no iico_ctrl).